tracer/src/Datadog.Trace/Debugger/Instrumentation/AsyncMethodDebuggerState.cs

```diff
@@ -0,0 +1,128 @@
+        /// Gets or sets a value indicating whether we should still be capturing values,
+        /// or halt for any reason (e.g an exception was caused by our instrumentation, rate limiter threshold reached).
+        /// </summary>
+        internal bool IsActive { get; set; } = true;
+
+        /// <summary>
+        /// Gets or sets a value indicating whether the current state of the async method captured locals or return value
+        /// </summary>
+        internal bool HasLocalsOrReturnValue { get; set; }
+
+        /// <summary>
+        /// Gets or sets a value indicating whether the current state of the async method captured arguments
+        /// </summary>
+        internal bool HasArguments { get; set; }
+
+        /// <summary>
+        /// Gets or sets a value of the parent (caller) state
+        /// </summary>
+        internal AsyncMethodDebuggerState Parent { get; set; }
+
+        /// <summary>
+        /// Gets the value of the MethodMetadataInfo that related to the current async method
+        /// </summary>
+        internal ref MethodMetadataInfo MethodMetadataInfo => ref MethodMetadataProvider.Get(MethodMetadataIndex);
+
+        /// <summary>
+        /// Gets the LiveDebugger SnapshotCreator
+        /// </summary>
+        internal DebuggerSnapshotCreator SnapshotCreator { get; }
+
+        /// <summary>
+        /// Gets or sets the LiveDebugger BeginMethod scope
+        /// </summary>
+        internal Scope Scope { get; set; }
+
+        /// <summary>
+        /// Gets or sets the LiveDebugger state StartTime
+        /// </summary>
+        internal DateTimeOffset? StartTime { get; set; }
+
+        /// <summary>
+        /// Gets or sets the Id of the probe
+        /// </summary>
+        internal string ProbeId { get; set; }
+
+        /// <summary>
+        /// Gets or sets the object that represents the "this" object of the async kick-off method (i.e. original method)
+        /// We can save it in some cases as TTarget but because we use it in serialization as boxed object anyway, we save it as object here
+        /// </summary>
+        public object KickoffInvocationTarget { get; set; }
+
+        /// <summary>
+        /// Gets or sets the object that represents the "this" object of the async MoveNext method
+        /// We can save it in some cases as TTarget but because we use it in serialization as boxed object anyway, we save it as object here
+        /// </summary>
+        public object MoveNextInvocationTarget { get; set; }
+
+        /// <summary>
+        /// Gets or sets the type that represents the "this" object of the async kick-off method (i.e. original method)
+        /// </summary>
+        public Type KickoffInvocationTargetType { get; set; }
+
+        /// <summary>
+        /// Gets or sets the type that represents the kickoff method of the state machine MoveNext (i.e. original method)
+        /// </summary>
+        public MethodBase KickoffMethod { get; set; }
```

Comment:
Nit: Shouldn't this be part of the `MethodMetadataInfo` object?
Two reasons why that might be preferable:

1. Seems more logically correct (current design seems to be that any piece of data whose lifetime is correlated to the specific method invocation resides inside the `[Async]MethodDebuggerState`, whereas any data that's about the method being invoked is in the `MethodMetadataInfo`.
2. This object is getting excessively large and in very extreme cases might have an impact on memory usage.

Reply:
Fixed in 5a103fbe57cedbc9e13568d4372c82a921c46e26